File: tools/mm/resources.py
#!/usr/bin/env python3
"""Sample landed testnet transactions per invocation category and collect,
for each: the DECLARED resources (the padded footprint and limits the client
signed: entries, instructions, bytes, resource fee) from the envelope, the
ACTUAL usage the host metered (`core_metrics` diagnostic events: cpu
instructions, entries and bytes read/written, events), and the fee actually
charged (`TransactionResult.fee_charged`).

Categories come from the maker/trader logs (tools/mm/mm.log, trader.log):
post-only places, heal walks (bucketed by phantom levels crossed), replaces,
batch replaces, settles, trader takes (bucketed by levels crossed), trader
rests. RPC retains about a day of history, so sampling walks each log
backwards from the end.

  python3 tools/mm/resources.py --contract C... [--per-cat 30] --out r.json
"""
import argparse
import collections
import concurrent.futures
import json
import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "soak"))
import soak  # noqa: E402


def categorize(path, kind):
    """Yield (category, txhash, meta) newest-first."""
    lines = open(path).read().splitlines()
    for line in reversed(lines):
        try:
            d = json.loads(line)
        except Exception:
            continue
        if d.get("outcome") != "ok" or not d.get("tx"):
            continue
        a = d["action"]
        if kind == "mm":
            if a == "place":
                yield "place post-only (maker quote)", d["tx"], {}
            elif a == "heal":
                p = d.get("phantoms") or 0
                b = "heal walk, 1-8 phantom levels" if p <= 8 else "heal walk, 9+ phantom levels"
                yield b, d["tx"], {"phantoms": p}
            elif a == "replace":
                yield "replace (single quote)", d["tx"], {}
            elif a == "replace_batch":
                yield "replace_batch (6-8 quotes)", d["tx"], {}
            elif a == "settle":
                yield "settle", d["tx"], {}
        else:
            if a == "take":
                c = d.get("crossed") or 0
                if c <= 1:
                    b = "place take, 0-1 levels crossed"
                elif c <= 3:
                    b = "place take, 2-3 levels crossed"
                else:
                    b = "place take, 4+ levels crossed"
                yield b, d["tx"], {"crossed": c, "lots": d.get("lots")}
            elif a == "rest":
                yield "place rest inside spread (trader)", d["tx"], {}
            elif a == "settle":
                yield "settle", d["tx"], {}


def fetch(cli, cat, h, meta):
    try:
        r = cli.rpc_call("getTransaction", {"hash": h})
        if r.get("status") != "SUCCESS":
            return None
        rc, out, _ = cli._run(["xdr", "decode", "--type", "TransactionEnvelope", "--output", "json"], stdin=r["envelopeXdr"])
        env = json.loads(out)
        tx = env["tx"]["tx"]
        sd = tx["ext"]["v1"]
        res = sd["resources"]
        fp = res["footprint"]
        rec = {
            "cat": cat, "tx": h, **meta,
            "d_ro": len(fp["read_only"]), "d_rw": len(fp["read_write"]),
            "d_instr": int(res["instructions"]), "d_read_b": int(res["disk_read_bytes"]), "d_write_b": int(res["write_bytes"]),
            "d_fee": int(sd["resource_fee"]), "tx_fee": int(tx["fee"]),
        }
        rc, out, _ = cli._run(["xdr", "decode", "--type", "TransactionResult", "--output", "json"], stdin=r["resultXdr"])
        rec["fee_charged"] = int(json.loads(out)["fee_charged"])
        for e in r.get("diagnosticEventsXdr") or []:
            rc, out, _ = cli._run(["xdr", "decode", "--type", "DiagnosticEvent", "--output", "json"], stdin=e)
            b = json.loads(out)["event"]["body"]["v0"]
            t = b["topics"]
            if len(t) == 2 and t[0].get("symbol") == "core_metrics":
                rec["a_" + t[1]["symbol"]] = int(b["data"]["u64"])
        return rec
    except Exception:
        return None


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--contract", required=True)
    ap.add_argument("--config-dir", default=".stellar")
    ap.add_argument("--network", default="testnet")
    ap.add_argument("--rpc", default="https://soroban-testnet.stellar.org")
    ap.add_argument("--mm-log", default="tools/mm/mm.log")
    ap.add_argument("--trader-log", default="tools/mm/trader.log")
    ap.add_argument("--per-cat", type=int, default=30)
    ap.add_argument("--out", default="resources.json")
    a = ap.parse_args()
    cli = soak.Cli(a.contract, a.config_dir, a.network, a.rpc)

    picked = collections.defaultdict(list)
    for path, kind in ((a.mm_log, "mm"), (a.trader_log, "trader")):
        if not os.path.exists(path):
            continue
        for cat, h, meta in categorize(path, kind):
            if len(picked[cat]) < a.per_cat:
                picked[cat].append((cat, h, meta))
    jobs = [j for v in picked.values() for j in v]
    print(f"{len(jobs)} transactions across {len(picked)} categories", file=sys.stderr)
    out = []
    with concurrent.futures.ThreadPoolExecutor(max_workers=8) as ex:
        futs = [ex.submit(fetch, cli, *j) for j in jobs]
        for i, f in enumerate(concurrent.futures.as_completed(futs)):
            r = f.result()
            if r:
                out.append(r)
            if (i + 1) % 25 == 0:
                print(f"{i+1}/{len(jobs)}", file=sys.stderr)
    with open(a.out, "w") as f:
        json.dump(out, f)
    print(f"wrote {len(out)} records to {a.out}", file=sys.stderr)


if __name__ == "__main__":
    main()

File: clients/web/ops/resources.ts
import { existsSync, readFileSync, writeFileSync } from "node:fs";
import { resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { createRpc, type Rpc } from "../src/book";
import { parseArgs, type ArgSpec } from "./lib/args";
import type { OpsLogRecord } from "./lib/logparse";
import { mapPool } from "./lib/pool";
import { decodeCoreMetrics, decodeEnvelopeResources, decodeFeeCharged } from "./lib/txdecode";

export type ResourcesArgs = {
  contract: string;
  configDir: string;
  network: string;
  rpc: string;
  mmLog: string;
  traderLog: string;
  perCat: number;
  out: string;
};

export const RESOURCES_SPECS: ArgSpec<keyof ResourcesArgs & string>[] = [
  { flag: "--contract", dest: "contract", required: true },
  { flag: "--config-dir", dest: "configDir", default: ".stellar" },
  { flag: "--network", dest: "network", default: "testnet" },
  { flag: "--rpc", dest: "rpc", default: "https://soroban-testnet.stellar.org" },
  { flag: "--mm-log", dest: "mmLog", default: "ops/mm.log" },
  { flag: "--trader-log", dest: "traderLog", default: "ops/trader.log" },
  { flag: "--per-cat", dest: "perCat", type: "int", default: 30 },
  { flag: "--out", dest: "out", default: "resources.json" },
];

export function parseResourcesArgs(argv: string[]): ResourcesArgs {
  return parseArgs<ResourcesArgs>(argv, RESOURCES_SPECS);
}

export type ResourcePick = { cat: string; tx: string; meta: Record<string, unknown> };

export function categorizeLine(d: OpsLogRecord, kind: "mm" | "trader"): ResourcePick | null {
  if (d.outcome !== "ok" || !d.tx) return null;
  const a = d.action ?? "";
  if (kind === "mm") {
    if (a === "place") return { cat: "place post-only (maker quote)", tx: String(d.tx), meta: {} };
    if (a === "heal") {
      const p = Number(d.phantoms ?? 0);
      const cat = p <= 8 ? "heal walk, 1-8 phantom levels" : "heal walk, 9+ phantom levels";
      return { cat, tx: String(d.tx), meta: { phantoms: p } };
    }
    if (a === "replace") return { cat: "replace (single quote)", tx: String(d.tx), meta: {} };
    if (a === "replace_batch") return { cat: "replace_batch (6-8 quotes)", tx: String(d.tx), meta: {} };
    if (a === "settle") return { cat: "settle", tx: String(d.tx), meta: {} };
    return null;
  }
  if (a === "take") {
    const c = Number(d.crossed ?? 0);
    let cat: string;
    if (c <= 1) cat = "place take, 0-1 levels crossed";
    else if (c <= 3) cat = "place take, 2-3 levels crossed";
    else cat = "place take, 4+ levels crossed";
    return { cat, tx: String(d.tx), meta: { crossed: c, lots: d.lots } };
  }
  if (a === "rest") return { cat: "place rest inside spread (trader)", tx: String(d.tx), meta: {} };
  if (a === "settle") return { cat: "settle", tx: String(d.tx), meta: {} };
  return null;
}

export function categorizeLog(text: string, kind: "mm" | "trader"): ResourcePick[] {
  const lines = text.split(/\r?\n/);
  const out: ResourcePick[] = [];
  for (let i = lines.length - 1; i >= 0; i--) {
    try {
      const d = JSON.parse(lines[i]) as OpsLogRecord;
      const pick = categorizeLine(d, kind);
      if (pick) out.push(pick);
    } catch {
      continue;
    }
  }
  return out;
}

export function pickPerCat(picks: ResourcePick[], perCat: number): ResourcePick[] {
  const picked = new Map<string, ResourcePick[]>();
  for (const p of picks) {
    const cur = picked.get(p.cat) ?? [];
    if (cur.length < perCat) {
      cur.push(p);
      picked.set(p.cat, cur);
    }
  }
  return [...picked.values()].flat();
}

export type ResourceRecord = {
  cat: string;
  tx: string;
  d_ro: number;
  d_rw: number;
  d_instr: number;
  d_read_b: number;
  d_write_b: number;
  d_fee: number;
  tx_fee: number;
  fee_charged: number;
  [k: string]: unknown;
};

export function recordFromTx(
  cat: string,
  hash: string,
  meta: Record<string, unknown>,
  tx: { status?: string; envelopeXdr?: string; resultXdr?: string; diagnosticEventsXdr?: string[] },
): ResourceRecord | null {
  if (tx.status && tx.status !== "SUCCESS") return null;
  if (!tx.envelopeXdr || !tx.resultXdr) return null;
  try {
    const decl = decodeEnvelopeResources(tx.envelopeXdr);
    const rec: ResourceRecord = {
      cat,
      tx: hash,
      ...meta,
      ...decl,
      fee_charged: decodeFeeCharged(tx.resultXdr),
      ...decodeCoreMetrics(tx.diagnosticEventsXdr ?? []),
    };
    return rec;
  } catch {
    return null;
  }
}

export type ResourcesDeps = {
  rpc?: Rpc;
  readFile?: (path: string) => string | null;
  exists?: (path: string) => boolean;
  write?: (path: string, data: string) => void;
};

export async function runResources(a: ResourcesArgs, deps: ResourcesDeps = {}): Promise<ResourceRecord[]> {
  const exists = deps.exists ?? existsSync;
  const readFile =
    deps.readFile ??
    ((path: string) => {
      try {
        return readFileSync(path, "utf8");
      } catch {
        return null;
      }
    });
  const write = deps.write ?? ((path, data) => writeFileSync(path, data));
  const rpc = deps.rpc ?? createRpc(a.rpc);

  const jobs: ResourcePick[] = [];
  for (const [path, kind] of [
    [a.mmLog, "mm"],
    [a.traderLog, "trader"],
  ] as const) {
    if (!exists(path)) continue;
    const text = readFile(path);
    if (text == null) continue;
    jobs.push(...pickPerCat(categorizeLog(text, kind), a.perCat));
  }
  process.stderr.write(`${jobs.length} transactions across ${new Set(jobs.map((j) => j.cat)).size} categories\n`);
  const out: ResourceRecord[] = [];
  let done = 0;
  const fetched = await mapPool(jobs, 8, async (j) => {
    try {
      const r = await rpc.getTransaction(j.tx);
      const rec = recordFromTx(j.cat, j.tx, j.meta, r);
      done += 1;
      if (done % 25 === 0) process.stderr.write(`${done}/${jobs.length}\n`);
      return rec;
    } catch {
      return null;
    }
  });
  for (const r of fetched) if (r) out.push(r);
  write(a.out, JSON.stringify(out));
  process.stderr.write(`wrote ${out.length} records to ${a.out}\n`);
  return out;
}

export async function main(argv = process.argv.slice(2)): Promise<void> {
  const a = parseResourcesArgs(argv);
  if (a.network !== "testnet") throw new Error("only testnet is supported by the ops entry points today");
  await runResources(a);
}

const entry = process.argv[1] ? resolve(process.argv[1]) : "";
if (entry && fileURLToPath(import.meta.url) === entry) {
  main().catch((e) => {
    process.stderr.write(String(e) + "\n");
    process.exit(1);
  });
}